### PR TITLE
Camera pose optimization 

### LINKED ISSUE #152: Enable optional pose optimization
Gradients can be optionally computed for the viewmat (only extrinsics) in the rasterizer backward pass but are currently not connected and cannot be enabled.

The goal should be to support viewmat gradients optionally.

It can be traced in the link below how to properly connect it
https://github.com/nerfstudio-project/gsplat/blob/961678f4819d909be60fdf8ee409acd3553be6e3/examples/simple_trainer.py#L149

### PR DESCRIPTION
**Pull request by calvin-laurenson**
Closes #152. Implements a direct delta method as well as a method using an MLP from [3R-GS](https://arxiv.org/pdf/2504.04294v1). As far as I can tell it works (deltas change a bit) but I don't really have data to confirm that. gsplat adds noise to poses and then tracks the l1 loss between the actual dataset poses and the optimized poses to benchmark. I might try that but I have a feeling it is going to be a lot more inconvenient in C++.

### CODE REVIEW

**Comment by MrNeRF** [comment]
Thank you. I will test it. The gsplat pose optimization is not doing too much from my experience. It only optimizes the extrinsics which is one reason. However, it is good to have something available at least.

**Comment by calvin-laurenson** [comment]
Should be good enough to merge.
--pose-opt none
| Scene | Iteration | PSNR | SSIM | LPIPS | Num Gaussians |
|-------|-----------|------|------|-------|---------------|
| garden | 30000 | 27.253824 | 0.837503 | 0.156080 | 1000000 |
| bicycle | 30000 | 25.407715 | 0.762797 | 0.253795 | 1000000 |
| stump | 30000 | 26.705208 | 0.801764 | 0.263052 | 1000000 |
| bonsai | 30000 | 32.434780 | 0.952602 | 0.249381 | 1000000 |
| counter | 30000 | 29.247484 | 0.930329 | 0.241964 | 1000000 |
| kitchen | 30000 | 31.471684 | 0.934473 | 0.155827 | 1000000 |
| room | 30000 | 32.326241 | 0.937071 | 0.272924 | 1000000 |

--pose-opt mlp
| Scene | Iteration | PSNR | SSIM | LPIPS | Num Gaussians |
|-------|-----------|------|------|-------|---------------|
| garden | 30000 | 25.576818 | 0.764070 | 0.169272 | 1000000 |
| bicycle | 30000 | 23.412964 | 0.678627 | 0.269298 | 1000000 |
| stump | 30000 | 25.680534 | 0.751829 | 0.272265 | 1000000 |
| bonsai | 30000 | 28.074051 | 0.878334 | 0.271342 | 1000000 |
| counter | 30000 | 23.351475 | 0.774738 | 0.309908 | 1000000 |
| kitchen | 30000 | 27.353262 | 0.836098 | 0.178690 | 1000000 |
| room | 30000 | 29.736109 | 0.905287 | 0.285291 | 1000000 |

Evaluation doesn't work obviously. Need to implement Procrustes analysis (See _Bundle-Adjusting Neural Radiance Field_ Section B.1). It errors out if you try to use pose opt and eval at the same time.

File: include/core/poseopt.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include "torch/nn/module.h"
+#include "torch/nn/modules/embedding.h"
+
+namespace gs {
+    struct PoseOptimizationModule : torch::nn::Module {
+        PoseOptimizationModule() {}
```
[the remark this answers is not in the document]
Comment (MrNeRF) [reply include/core/poseopt.hpp:7]: This seems critical at a first sight?

[the remark this answers is not in the document]
Comment (calvin-laurenson) [reply include/core/poseopt.hpp:7]: I don't think so. PoseOptimizationModule doesn't have any parameters on purpose; it is just a null module that returns unchanged extrinsics. I'm not very familiar with the torch cpp frontend but I have never seen the copilot recommended code in examples.

File: src/poseopt.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    DirectPoseOptimizationModule::DirectPoseOptimizationModule(int number_of_cameras)
+        : camera_embeddings(register_module("camera_embeddings",
+                                            torch::nn::Embedding(number_of_cameras, 9))), rot_identity(register_buffer(
+                                                "rot_identity",
+                                                torch::tensor({1.0f, 0.0f, 0.0f, 0.0f, 1.0f, 0.0f}))) {
+        torch::nn::init::zeros_(camera_embeddings->weight);
+    }
+    // Only supports 1d batch size
+    torch::Tensor DirectPoseOptimizationModule::forward(torch::Tensor camera_transforms, torch::Tensor embedding_ids) {
+        auto bs = camera_transforms.size(0);
```
[the remark this answers is not in the document]
Comment (calvin-laurenson) [reply src/poseopt.cpp:28]: this suggestion is just wrong. there is a difference between 1d and batch and a single item batch

File: src/trainer.cpp [hunk 3]
```diff
@@ -156,6 +158,26 @@ namespace gs {
...
+            }
+            if (params.optimization.pose_optimization == "direct") {
+                poseopt_module_ = std::make_unique<gs::DirectPoseOptimizationModule>(train_dataset_->get_cameras().size());
+            } else if (params.optimization.pose_optimization == "mlp") {
+                poseopt_module_ = std::make_unique<gs::MLPPoseOptimizationModule>(train_dataset_->get_cameras().size());
+            } else {
+                throw std::runtime_error("Invalid pose optimization type: " + params.optimization.pose_optimization);
+            }
+            poseopt_optimizer_ = std::make_unique<torch::optim::Adam>(
+                std::vector<torch::Tensor>{poseopt_module_->parameters()},
```
[the remark this answers is not in the document]
Comment (calvin-laurenson) [reply src/trainer.cpp:168]: the optimizer is only create if the pose opt type is not none (none is the only case where PoseOptimizationModule is directly created)

File: src/camera.cpp [hunk 4]
```diff
@@ -43,7 +43,23 @@ namespace gs {
           _image_height(camera_height),
           _world_view_transform{world_to_view(R, T)} {
     }
-
+    Camera::Camera(const Camera& other, const torch::Tensor& transform)
```
[the remark this answers is not in the document]
Comment (calvin-laurenson) [reply src/camera.cpp:53]: I don't really like that I had to make this constructor but I couldn't think of a better way to make it work.

**Comment by calvin-laurenson** [comment]
Blocked by #294

**Comment by calvin-laurenson** [comment]
I implemented gradients for `w2c` based on https://github.com/nerfstudio-project/gsplat/pull/127. I'm not very confident with my CUDA code but training doesn't blow up and there are some pose adjustments being made.

**Comment by MrNeRF** [comment]
Thx, I will check.

**Comment by MrNeRF** [comment]
Left without pose optim, right with pose-optim mlp!

https://github.com/user-attachments/assets/557e9e0d-1168-49d2-a230-62c6243d1540
